### ISSUE TITLE
Not maintained anymore?

### ISSUE THREAD

**Issue by dmitry** [reactql/kit#147]
@leebenson are you still using ReactQL in your daily routine or are you going to continue to maintain it in the nearest future?

**Comment by leebenson** [comment]
I'm personally still using ReactQL in production, including on a fairly high-trafficked e-commerce page that's taking live orders.

There are a few outstanding PRs and issues I'd like to get to; it's not been updated in a little while simply because I haven't had a pressing use-case to bump any of the libs, and I've been a [little busy elsewhere](https://github.com/reactql/kit/issues/110#issuecomment-355864962)

I'm kinda blocking on [this thread](https://github.com/reactql/kit/issues/80), which are thoughts around where to go from here with Apollo 2, converting to Typescript, etc. I've flip-flopped on approaches, mostly because many of the libs I'd be building it on (Apollo 2, possibly Webpack v4) are somewhat in a state of flux and/or don't necessarily have ready patterns to replacing the parts of the stack that would make any significant upgrade to ReactQL worthwhile. And also because the impetus to switch just hasn't been there on any personal projects, because the current versions work perfectly well.

I'll probably get to it eventually. In the meantime - and as always - I recommend that anyone using ReactQL do as more as a starter kit, rather than relying on it as a framework that's relied upon to keep track with specific third-party lib versions. No doubt you'll find that your own projects will eventually have unique requirements ReactQL probably doesn't address anyway, so it's best not to rely on this (or any) starter kit to keep track of any particular lib.

**Comment by birdiesanders** [comment]
I'm not sure if my input counts for much, but I've been using your framework as a base for a company management platform, and the way that you have things working now works out pretty damned well. I don't use redux in it as of yet, but the way everything is set up is just fantastic. Absolutely great work gluing it all together. Converting to typescript means I would have to teach everyone a new dialect of javascript that they will only use when dealing with this one system, and that isn't so optimal. The main reason to upgrade apollo is to be able to use the Apollo dev tools, which are pretty handy once you have 10 mutations and a dozen queries that the apollo redux store is tracking, and it's better than the Redux dev tools, so that would be worth the effort in at least that regard, even if you left everything else as is, maybe upgrade to webpack 4 so we can squeeze it a bit harder, but there will be helpers that come out in the coming months to ease the transition to that.

**Comment by birdiesanders** [comment]
Oh wow! Congrats on the baby! They are freaking amazing.

**Comment by leebenson** [comment]
Thanks @birdiesanders, appreciate your input and kind words.

I'm working on some Webpack 4 stuff locally, and so far it's going well. The ecosystem around it is slowly building, and many plugins have _next_ versions in alpha/beta that are compatible.

I'm also experimenting with proper async code-splitting -- e.g. loading the relevant chunks in the SSR for JS and CSS, rather than rendering HTML and then having the client re-request the chunk. AFAIK, there's no practical example of that working with Webpack 4 yet... so I'm pretty excited about that.

Re: Typescript, the point is to enable Intellisense for a growing config API. It'll be 'opt-in' outside of the core kit- your userland code can remain as plain ES6 JS and you'll never need to think about TS, unless you're delving into the kit itself. 90% of the time that won't be required, due to the flexibility of userland config. Hopefully this is a good middle ground for most teams.

There's still a bit of exploring for me to do with Apollo 2 before I'd be comfortable swapping out Redux entirely, but concepts/plugins like local link state will eventually solve state management entirely.

I'll get to it soon. In the meantime, my advice to anyone would be: don't wait to build whatever you're working on. The current version is probably good enough, and everything that the next kit will bring is mostly additive / building on what's already there.

**Comment by iamclaytonray** [comment]
@leebenson - Do you want some help maintaining? I'm more than happy to contribute. I personally just found out about this lib so not sure of all the inner-workings but I can adapt. Let me know!

**Comment by leebenson** [comment]
@iamclaytonray - thanks for the offer, appreciated. I'm actually now pretty close to releasing a new version of ReactQL, which will bring about some significant changes. I'd suggest waiting for that to land first, so you'll be able to assess whether you'd like to contribute to the core, or use at as a base for a forked version.

I'll announce details in this issue once it lands.

**Comment by leebenson** [comment]
To update this thread: I've committed the first alpha version of Launch.js, a micro-framework that will become the foundation of ReactQL v3:

https://github.com/launchjs/app

The premise is that it abstracts away Webpack and server configuration, allowing configs to built on top of it. The Koa server is instantiated _before_ Webpack, but then passed _to_ Webpack, allowing a clean separation between 'boilerplate' and 'userland' code, yet with access to the non-Webpack'd server from within your Webpack'd code.

This makes it trivial to do things such as refer back to Webpack stats from within your built assets, from adding the correct client chunks to your SSR HTML `<script>` tags, to accessing Koa middleware defined outside of Webpack -- the configuration options become unlimited.

There are a few extra niceties for ReactQL, such as hot-module reloading on the same port (on both the server, and in-browser) and being able to import ReactQL as a single NPM module and avoid all of the boilerplate, yet have the same level fo configurability as you're used to in v2.

It uses Webpack 4, which also builds significantly faster and has more efficient chunking of user/vendor code on the client.

And it's written in Typescript (but also accepts plain `.js` userland code), so you get intellisense everywhere:

<img width="848" alt="screen shot 2018-04-03 at 09 33 39" src="https://user-images.githubusercontent.com/719722/38238567-226a2b3c-3722-11e8-9cd8-dad58da4de17.png">

This is an early alpha release, so there's still a few things left to do before it's ready to build on. Most of those things should be taken care of this week.

The next version of ReactQL will transition to Launch.js and will bump to Apollo 2 and React 16.3. It'll also drop Redux and use GraphQL for everything.

Stay tuned 👍

**Comment by mnpenner** [comment]
Thanks for all the hard work @leebenson! Just wanted to say don't feel rushed to push out Apollo 2 if you don't think it's ready for prime-time yet. I know I've been bitten by alphas, betas, and 2.0s plenty of times.

The claims about Webpack 4 being up to 95% faster don't appear to be true either. I've seen zero improvement in my main project, and extract-text-webpack-plugin stopped working. I've now switched it out for mini-css-extract-plugin and added in DLL plugin to regain some speed (huge difference BTW), but net gain has been zilch. 

I know it's important to keep up with releases, but feel free to a lag behind a *little* bit so they can stabilise :smile: 

On another note, have you done any research into cache-busting? I've been having a heck of a time trying to get webpack to *not* change all the bundle hashes every time I edit some code. If you release new versions of your web-app every other day, forcing all your clients to redownload *everything* is not ideal.

DllPlugin seems to help alleviate this by putting the vendor stuff into a *completely* separate bundle, but otherwise, any `import()` seems to affect all the chunks (HashedModuleIds or not). Webpack doesn't support content hashes, only chunk hashes AFAICT.

**Comment by leebenson** [comment]
Thanks @mnpenner, an astute analysis, as always!

I completely agree that tracking edge releases just for the sake of 'keeping up', is pretty pointless. I don't want to add to JS fatigue unnecessarily - the API changes between major versions of dependent libraries often mean un/relearning concepts, so the gain has to be worth dragging users along for the ride. At the same time, I'm taking a fundamentally new approach with Launch.js, so it makes sense to start afresh with current libs rather than just rehash what v1 already offers.

IMO, Apollo 2 warrants a switch. The new API is more composable, and their dev team is very active in releasing useful additions. Plus all the docs on their site have been updated to reference v2, and since that client is pretty central to why ReactQL exists, I think it's important that users feel a) they're using the best of what Apollo has to offer, and b) they can easily find answers without resorting to digging up archived docs. Like anything in the JS world, the approach/concepts are a moving target, but I think a lot of it is ultimately making GraphQL easier to design for.

Webpack 4 is a bit more of a wildcard, because it's an ecosystem, meaning a more disjointed experience with dependent loaders. I've been down the Extract-Text/Mini-CSS road, and the same again for other loaders that weren't quite ready for primetime. A lot of progress seems to have been made in the last month with many of the more popular loaders to bring their plugin hooks up to date with the v4 spec. We seem to be over the worst of it, although Extract-Text is basically saying 'don't use this anymore' and Mini-CSS still isn't _quite_ there, yet.

FWIW, in Launch.js, CSS/style config will be a separate package that can be optionally loaded. I know many prefer a pure JS solution to styles these days, and in some cases that makes it easier to inject above-the-fold rather than fighting with CSS modules (although the whole point of Mini-CSS is to be able to chunk assets in a similar fashion to Webpack's handling of JS modules.) I've yet to tackle styles in this new approach; I'll run into these issues in the days ahead, no doubt.

Despite the issues that come with early adoption, I was attracted to Webpack 4 for mainly two reasons:

1. Compilation speed (at least for Typescript) seems to have improved quite dramatically, without resorting to HappyPack/multi-threading via plugins; Launch.js spins up a test build from cold in ~1.2s, even before targeting any caching/DLL strategies, and greatly simplifies the config.

2. Chunking is vastly improved- no more dumping everything under `node_modules` into a single `vendor.js` and loading the whole lot with every request. That's pretty substantial for projects that tend to use heavier modules on select routes, and don't want the overhead of loading everything at once. Combined with some custom code in Launch.js for server awareness over which client chunks to output in the initial HTML of any given route (TBA), and code-splitting should be a massively upgraded experience. (Async imports directly benefit, too, since chunking/splitting are basically two sides of the same coin.)

Re: cache busting, that's not something I've spent a lot of time on so far. By default, assets in Launch.js are built with a `[name].[chunkhash].js` format, which should (in theory) mean that as long as the content in the chunk doesn't change, neither does the hash. There's an [official Webpack example here](https://github.com/webpack/webpack/tree/master/examples/chunkhash) that offers some discussion:

> A common challenge with combining [chunkhash] and Code Splitting is that the entry chunk includes the webpack runtime and with it the chunkhash mappings. This means it's always updated and the [chunkhash] is pretty useless, because this chunk won't be cached.
>
> A very simple solution to this problem is to create another chunk which contains only the webpack runtime (including chunkhash map). This can be achieved with the optimization.runtimeChunk options. To avoid the additional request for another chunk, this pretty small chunk can be inlined into the HTML page.

This is something I'll explore further (along with the DLL Plugin), to hopefully come up with some sensible defaults for caching re: filenames, and build speed. I'll take a deeper look once a stable release of Launch.js is out the door. Should only be a few days for that, btw -- I'm working on firming up 'dev' mode, then I'll put together the first version of the ReactQL 'preset' for user feedback.

**Comment by leebenson** [comment]
FWIW, I've released v1.0 of the CSS plugin for Launch.js at https://github.com/launchjs/styles, if you're curious how adding plugins (like ReactQL itself) will work.

[This is pertinent source](https://github.com/launchjs/styles/blob/master/src/styles.ts#L108-L225) that illustrates how merging Webpack configs will work (in this case, Mini-CSS-Extract-Plugin) on the client/server.

It's still not quite there in terms of loading _just_ the chunks that pertain to any given route, but it has parity with the current version in ReactQL in terms of it providing an isomorphic, hot-reloading styling experience in development.

And this time, server and client modes run from the same port, and get reloaded automatically on changes.

Now to finish the ReactQL plugin...

**Comment by aoighost** [comment]
I'm getting ready to start a new app based on this. Should I wait, if so, how long do you estimate until the new starter pack is up?

**Comment by leebenson** [comment]
@aoighost:

> In the meantime - and as always - I recommend that anyone using ReactQL do so more as a starter kit, rather than relying on it as a framework that's expected to keep track with specific third-party lib versions. No doubt you'll find that your own projects will eventually have unique requirements ReactQL probably doesn't address anyway, so it's best not to rely on this (or any) starter kit to maintain version parity with third-party modules.

Don't wait for me. Evaluate if ReactQL is good enough to get you moving, or roll your own requirements. As someone that has put too many projects on hold waiting for a big, shiny new library to land... trust me on this. Use the tools you have today.

**Comment by CaptainChemist** [comment]
@aoighost Maybe open up a new issue? I use reactql for several projects and I think it works great. Just make sure you are using a new version of npm/node or you get build errors. I use nvm to set node to 8.4 and that works just fine, but anything new enough should be good.

**Comment by leebenson** [comment]
We're back! https://github.com/leebenson/reactql
